File: Sources/PUPSISPortalApp/Core/DayAgenda.swift
import Foundation

/// Where a class sits relative to right now.
enum ClassPhase {
    case past
    case inSession
    case upcoming
}

/// "What does today look like right now" — the shared reading behind both the
/// Today screen and the menu bar. Pure on purpose: no `Preferences`, no views,
/// no clock of its own. Vacancy arrives as a closure taking the class **and its
/// occurrence date**, so each caller applies its own per-week rule and tomorrow
/// keys to the right week (which can be next week) without this type depending
/// on `Preferences`.
struct DayAgenda {
    struct Item: Identifiable {
        let session: ClassSession
        let phase: ClassPhase
        var id: String { session.id }
    }

    /// Today's non-vacant classes, sorted by start, each tagged with its phase.
    let items: [Item]
    /// The first non-vacant class tomorrow, or `nil`.
    let tomorrowFirst: ClassSession?

    static func make(
        sessions: [ClassSession],
        now: Date,
        isVacant: (ClassSession, Date) -> Bool,
        calendar: Calendar = .current
    ) -> DayAgenda {
        let today = Weekday.on(now, calendar: calendar)
        let nowMinutes = NowLine.minutes(of: now, calendar: calendar)

        let items = sessions
            .filter { $0.day == today && !isVacant($0, now) }
            .sorted { $0.start < $1.start }
            .map { session -> Item in
                let phase: ClassPhase
                if session.end <= nowMinutes { phase = .past }
                else if session.start <= nowMinutes { phase = .inSession }
                else { phase = .upcoming }
                return Item(session: session, phase: phase)
            }

        let tomorrowDate = calendar.date(byAdding: .day, value: 1, to: now) ?? now
        let tomorrowDay = Weekday.on(tomorrowDate, calendar: calendar)
        let tomorrowFirst = sessions
            .filter { $0.day == tomorrowDay && !isVacant($0, tomorrowDate) }
            .min { $0.start < $1.start }

        return DayAgenda(items: items, tomorrowFirst: tomorrowFirst)
    }

    /// Today's classes that haven't finished — what the menu bar shows, so the
    /// scarce space reads as "now and what's left", not the whole past day.
    var remaining: [Item] {
        items.filter { $0.phase != .past }
    }

    // MARK: Merged timeline (classes + custom calendar events)

    /// A single row in the Today timeline — a class or a calendar event, flattened
    /// to the same shape so free time can be read across both.
    struct AgendaEntry: Identifiable {
        enum Kind: Equatable {
            case klass(ClassSession)
            case event
        }
        let id: String
        /// Minutes from midnight.
        let start: Int
        let end: Int
        let title: String
        let subtitle: String
        let kind: Kind
        let phase: ClassPhase

        var session: ClassSession? {
            if case .klass(let session) = kind { return session }
            return nil
        }
    }

    /// Today's non-vacant classes merged with today's calendar `events`, sorted by
    /// start and phase-tagged — so the Today screen can draw one timeline and its
    /// free stretches span classes *and* the user's own events, not just classes.
    static func timeline(
        classes: [ClassSession],
        events: [DayBlock],
        now: Date,
        isVacant: (ClassSession, Date) -> Bool,
        calendar: Calendar = .current
    ) -> [AgendaEntry] {
        let today = Weekday.on(now, calendar: calendar)
        let nowMinutes = NowLine.minutes(of: now, calendar: calendar)

        func phase(start: Int, end: Int) -> ClassPhase {
            if end <= nowMinutes { return .past }
            if start <= nowMinutes { return .inSession }
            return .upcoming
        }

        let classEntries = classes
            .filter { $0.day == today && !isVacant($0, now) }
            .map { session in
                AgendaEntry(
                    id: "class-\(session.id)",
                    start: session.start, end: session.end,
                    title: session.subjectCode, subtitle: session.description,
                    kind: .klass(session),
                    phase: phase(start: session.start, end: session.end)
                )
            }

        let eventEntries = events
            .filter { $0.day == today }
            .map { block in
                AgendaEntry(
                    id: block.id,
                    start: block.start, end: block.end,
                    title: block.title, subtitle: block.subtitle,
                    kind: .event,
                    phase: phase(start: block.start, end: block.end)
                )
            }

        return (classEntries + eventEntries).sorted { $0.start < $1.start }
    }

    /// Free minutes still ahead today — the ≥15-minute gaps the timeline draws
    /// between consecutive entries, counting only gaps that haven't passed.
    /// Overlapping entries contribute nothing (a negative gap is dropped).
    static func remainingFreeMinutes(_ entries: [AgendaEntry], nowMinutes: Int) -> Int {
        var total = 0
        for i in entries.indices.dropLast() {
            let gap = entries[i + 1].start - entries[i].end
            guard gap >= 15, entries[i + 1].start > nowMinutes else { continue }
            total += gap
        }
        return total
    }
}

File: windows/PUPSISPortal.Core/DayAgenda.cs
namespace PUPSISPortal.Core;

/// <summary>
/// Where a class sits relative to right now.
/// </summary>
public enum ClassPhase
{
    Past,
    InSession,
    Upcoming,
}

/// <summary>
/// "What does today look like right now" — the shared reading behind both the
/// Today screen and the menu bar. Pure on purpose: no Preferences, no views,
/// no clock of its own. Vacancy arrives as a closure taking the class and its
/// occurrence date, so each caller applies its own per-week rule and tomorrow
/// keys to the right week (which can be next week) without this type depending
/// on Preferences.
/// </summary>
public record DayAgenda
{
    public record Item(ClassSession Session, ClassPhase Phase)
    {
        public string Id => Session.Id;
    }

    /// <summary>
    /// Today's non-vacant classes, sorted by start, each tagged with its phase.
    /// </summary>
    public required List<Item> Items { get; set; }

    /// <summary>
    /// The first non-vacant class tomorrow, or null.
    /// </summary>
    public ClassSession? TomorrowFirst { get; set; }

    /// <summary>
    /// Today's classes that haven't finished — what the menu bar shows, so the
    /// scarce space reads as "now and what's left", not the whole past day.
    /// </summary>
    public List<Item> Remaining => Items.Where(i => i.Phase != ClassPhase.Past).ToList();

    public static DayAgenda Make(
        IEnumerable<ClassSession> sessions,
        DateTime now,
        Func<ClassSession, DateTime, bool> isVacant)
    {
        var today = WeekdayExtensions.On(now);
        var nowMinutes = NowLine.Minutes(now);

        var sessionsList = sessions.ToList();
        var items = sessionsList
            .Where(s => s.Day == today && !isVacant(s, now))
            .OrderBy(s => s.Start)
            .Select(session =>
            {
                var phase = session.End <= nowMinutes ? ClassPhase.Past
                    : session.Start <= nowMinutes ? ClassPhase.InSession
                    : ClassPhase.Upcoming;
                return new Item(session, phase);
            })
            .ToList();

        var tomorrowDate = now.AddDays(1);
        var tomorrowDay = WeekdayExtensions.On(tomorrowDate);
        var tomorrowFirst = sessionsList
            .Where(s => s.Day == tomorrowDay && !isVacant(s, tomorrowDate))
            .OrderBy(s => s.Start)
            .FirstOrDefault();

        return new DayAgenda { Items = items, TomorrowFirst = tomorrowFirst };
    }

    // MARK: Merged timeline (classes + custom calendar events)

    /// <summary>
    /// A single row in the Today timeline — a class or a calendar event, flattened
    /// to the same shape so free time can be read across both.
    /// </summary>
    public record AgendaEntry
    {
        public enum KindType { Klass, Event }

        public required string Id { get; set; }
        /// <summary>
        /// Minutes from midnight.
        /// </summary>
        public required int Start { get; set; }
        public required int End { get; set; }
        public required string Title { get; set; }
        public required string Subtitle { get; set; }
        public required KindType Kind { get; set; }
        public required ClassPhase Phase { get; set; }

        public ClassSession? Session { get; set; }
    }

    /// <summary>
    /// Today's non-vacant classes merged with today's calendar events, sorted by
    /// start and phase-tagged — so the Today screen can draw one timeline and its
    /// free stretches span classes and the user's own events, not just classes.
    /// </summary>
    public static List<AgendaEntry> Timeline(
        IEnumerable<ClassSession> classes,
        IEnumerable<DayBlock> events,
        DateTime now,
        Func<ClassSession, DateTime, bool> isVacant)
    {
        var today = WeekdayExtensions.On(now);
        var nowMinutes = NowLine.Minutes(now);

        ClassPhase Phase(int start, int end) =>
            end <= nowMinutes ? ClassPhase.Past
            : start <= nowMinutes ? ClassPhase.InSession
            : ClassPhase.Upcoming;

        var classList = classes.ToList();
        var classEntries = classList
            .Where(s => s.Day == today && !isVacant(s, now))
            .Select(session => new AgendaEntry
            {
                Id = $"class-{session.Id}",
                Start = session.Start,
                End = session.End,
                Title = session.SubjectCode,
                Subtitle = session.Description,
                Kind = AgendaEntry.KindType.Klass,
                Phase = Phase(session.Start, session.End),
                Session = session,
            })
            .ToList();

        var eventList = events.ToList();
        var eventEntries = eventList
            .Where(b => b.Day == today)
            .Select(block => new AgendaEntry
            {
                Id = block.Id,
                Start = block.Start,
                End = block.End,
                Title = block.Title,
                Subtitle = block.Subtitle,
                Kind = AgendaEntry.KindType.Event,
                Phase = Phase(block.Start, block.End),
            })
            .ToList();

        return (classEntries.Concat(eventEntries).OrderBy(e => e.Start)).ToList();
    }

    /// <summary>
    /// Free minutes still ahead today — the ≥15-minute gaps the timeline draws
    /// between consecutive entries, counting only gaps that haven't passed.
    /// Overlapping entries contribute nothing (a negative gap is dropped).
    /// </summary>
    public static int RemainingFreeMinutes(IEnumerable<AgendaEntry> entries, int nowMinutes)
    {
        var entryList = entries.ToList();
        var total = 0;

        for (int i = 0; i < entryList.Count - 1; i++)
        {
            var gap = entryList[i + 1].Start - entryList[i].End;
            if (gap >= 15 && entryList[i + 1].Start > nowMinutes)
            {
                total += gap;
            }
        }

        return total;
    }

    public DayAgenda() { }
}

/// <summary>
/// Helper to extract minutes from midnight from a DateTime.
/// </summary>
public static class NowLine
{
    public static int Minutes(DateTime now)
    {
        return now.Hour * 60 + now.Minute;
    }
}
